File: 17_Math_&_Geometry/08_Detect_Squares/2013-detect-squares.cpp
/*
Problem: LeetCode 2013 - Detect Squares

Intuition:
Precompute the count of points in each row and column and then use this information to find the other three points of the square.

Approach:
1. We use two unordered maps "rowMap" and "colMap" to keep track of the count of points in each row and column, respectively.
2. When we add a new point (x, y), we increment its count in both "rowMap" and "colMap" accordingly.
3. To count the number of squares with a given point (x, y) as the bottom-right corner, we look for other points in the same row "x."
4. For each point found at column "col" in row "x," we calculate the edge length of the potential square. Let's call it "edgeLen," which is the absolute difference between "col" and "y."
5. Then, we check if the points (x - edgeLen, col), (x + edgeLen, col), and (x + edgeLen, y) exist in the maps. If they do, we have found a square with (x, y) as the bottom-right corner.
6. We repeat this process for all points in row "x" to find all squares with (x, y) as the bottom-right corner.
7. Finally, we sum up the counts of all these squares to get the total number of squares with (x, y) as the bottom-right corner.
8. The process is efficient as we use unordered maps to store the counts, making the lookups quick and allowing us to count squares in constant time.

Time Complexity:
The time complexity of this approach is O(1) for `add` and O(n) for `count` (where "n" is the average number of points in the same row as the given point).

Space Complexity:
The space complexity is O(N) (where "N" is the total number of points added).
*/

class DetectSquares {
  public:
    // We maintain two maps for row and column coordinates to store the count of points in each row and column.
    vector<unordered_map<int, int>> rowMap;
    vector<unordered_map<int, int>> colMap;

    DetectSquares() : rowMap(1001), colMap(1001) {
        // Constructor to initialize the rowMap and colMap with size 1001.
    }

    // Function to add a point to the grid.
    void add(vector<int> point) {
        // Increment the count of points in the corresponding row and column maps.
        rowMap[point[0]][point[1]]++;
        colMap[point[1]][point[0]]++;
    }

    // Function to count the number of squares that can be formed with a given point as the bottom-right corner.
    int count(vector<int> point) {
        int result = 0;

        // Iterate through the row map for the given point's row.
        for (auto [col, cCount] : rowMap[point[0]]) {
            int edgeLen = abs(col - point[1]);

            if (edgeLen == 0) {
                continue;    // Skip points that are at the same position as the given point.
            }

            // Calculate the row coordinates for the other two points of the square based on the edge length.
            int row = point[0] - edgeLen;

            if (colMap[col].find(row) != colMap[col].end()) {
                auto rCount = colMap[col][row];

                // Check if the other two points exist in the colMap.
                if (colMap[point[1]].find(row) != colMap[point[1]].end()) {
                    // If so, update the result by multiplying the counts of the four points.
                    result += (rCount * cCount * colMap[point[1]][row]);
                }
            }

            row = point[0] + edgeLen;

            if (colMap[col].find(row) != colMap[col].end()) {
                auto rCount = colMap[col][row];

                // Check if the other two points exist in the colMap.
                if (colMap[point[1]].find(row) != colMap[point[1]].end()) {
                    // If so, update the result by multiplying the counts of the four points.
                    result += (rCount * cCount * colMap[point[1]][row]);
                }
            }
        }

        return result;
    }
};

File: 17_Math_&_Geometry/08_Detect_Squares/2013-detect-squares.rs
/*
Problem: LeetCode 2013 - Detect Squares

Key Idea:
A rectangle requires two pairs of points to share one common point and have equal distances along both axes.

Approach:
1. Create a `DetectSquares` struct to maintain a `points_count` HashMap, which counts the occurrences of each point.
2. In the `add` method, update the count for the given point in the `points_count` HashMap.
3. In the `count` method, iterate through the points in the `points_count` HashMap.
4. For each point `key`, check if it can form a rectangle with the input point `point`. If so, calculate the count of two additional points to complete the rectangle.
5. Multiply these three counts (including `val`, which represents the count of `key`) to get the total count of rectangles formed with `point` as one corner.
6. Add this count to the result and continue iterating.
7. Finally, return the result, which represents the total count of rectangles formed by the given points.

Time Complexity:
O(n), due to count() method.

Space Complexity:
O(n), for using a hashmap.
*/

use std::collections::HashMap;

struct DetectSquares {
    points_count: HashMap<Vec<i32>, i32>,
}

impl DetectSquares {
    fn new() -> Self {
        DetectSquares {
            points_count: HashMap::new(),
        }
    }

    fn add(&mut self, point: Vec<i32>) {
        self.points_count
            .entry(point)
            .and_modify(|count| *count += 1)
            .or_insert(1);
    }

    fn count(&self, point: Vec<i32>) -> i32 {
        let mut result: i32 = 0;

        for (key, val) in self.points_count.iter() {
            if key[0] != point[0]
                && key[1] != point[1]
                && (point[0] - key[0]).abs() == (point[1] - key[1]).abs()
            {
                let count_a = match self.points_count.get(&*vec![key[0], point[1]]) {
                    Some(&count) => count,
                    None => 0,
                };

                let count_b = match self.points_count.get(&*vec![point[0], key[1]]) {
                    Some(&count) => count,
                    None => 0,
                };

                result += val * count_a * count_b;
            }
        }

        result
    }
}
